### ISSUE TITLE
Normal files on the prompt should have `fi` rather than `rs` $LS_COLORS type

### ISSUE THREAD

**Issue by qwenger** [xonsh/xonsh#3608]
I'm currently reworking colors and styles in my shell and came across $LS_COLORS. Trying to get an overview of all values, I compiled the following table from several resources: (x = mentioned)

| | | [1] | [2] | [3] | xonsh [4] | my own `dircolors --print-database` |
|---|---|---|---|---|---|---|
| no | NORMAL, NORM				| x  |     |  x  |     |  x  |
| fi | FILE						| x  |     |  x  |     |  x  |
| di | DIR						| x  |  x  |  x  |  x  |  x  |
| ln | SYMLINK, LINK, LNK		| x  |  x  |  x  |  x  |  x  |
| pi | FIFO, PIPE				| x  |  x  |  x  |  x  |  x  |
| do | DOOR						| x  |  x  |  x  |  x  |  x  |
| bd | BLOCK, BLK				| x  |  x  |  x  |  x  |  x  |
| cd | CHAR, CHR				| x  |  x  |  x  |  x  |  x  |
| or | ORPHAN					| x  |  x  |  x  |  x  |  x  |
| so | SOCK						| x  |  x  |  x  |  x  |  x  |
| su | SETUID					| x  |  x  |  x  |  x  |  x  |
| sg | SETGID					| x  |  x  |  x  |  x  |  x  |
| tw | STICKY_OTHER_WRITABLE	| x  |  x  |  x  |  x  |  x  |
| ow | OTHER_WRITABLE			| x  |  x  |  x  |  x  |  x  |
| st | STICKY					| x  |  x  |  x  |  x  |  x  |
| ex | EXEC						| x  |  x  |  x  |  x  |  x  |
| mi | MISSING					| x  |  x  |  x  |  x  |  x  |
| lc | LEFTCODE, LEFT			| x  |     |     |     |     |
| rc | RIGHTCODE, RIGHT			| x  |     |     |     |     |
| ec | ENDCODE, END				| x  |     |     |     |     |
| rs | RESET					|    |  x  |     |  x  |  x  |  
| mh | MULTIHARDLINK			|    |  x  |     |  x  |  x  |
| ca | CAPABILITY				|    |  x  |     |  x  |  x  |

xonsh's support for types seems good overall. Let's concentrate on the problematic details:

1. `lc`, `rc`, `ec` look like special types for other shells, so they probably can be safely ignored. `no` probably falls in this category, too.
2. Stranger is the missing `fi`. Looking at [5], it seems that `rs` is a fallback on xonsh's prompt (which could make sense in some cases). [6] seems to imply that `rs` would be "regular", but it is actually a reset code (see the table above).
3. [5] informs about unhandled `ca` in the docstring. It says nothing about `mh`, which it does not support either.

## TL;DR: Here is what I think is problematic at the moment:

1. Normal files should get assigned `fi`, not `rs` in [5] and [6]. In the current situation, they are colored differently in `ls` than on the prompt, at least for me.
2. [5] should probably either support `mh` or its lack of support be noted in the docstring. (but TBH I'm not really sure what `mh` is about) (edit: see the edit5 below)
3. This is probably a separate issue, but types that define a background color in [4] make their background color spill on the following lines until the next prompt. (edit: see the edit3 below)

[1] http://www.bigsoft.co.uk/blog/2008/04/11/configuring-ls_colors
[2] https://www.networkworld.com/article/3138909/coloring-your-world-with-ls-colors.html
[3] https://askubuntu.com/questions/466198/how-do-i-change-the-color-for-directories-with-ls-in-the-console
[4] https://github.com/xonsh/xonsh/blob/master/xonsh/environ.py#L199
[5] https://github.com/xonsh/xonsh/blob/master/xonsh/pyghooks.py#L1359
[6] https://github.com/xonsh/xonsh/blob/master/tests/test_pyghooks.py#L160

## EDIT: steps to reproduce:

1. Add in your .xonshrc:

```
$LS_COLORS.update({
     "fi": ("BOLD_GREEN",),
     "rs": ("BOLD_RED",),
     "st": ("BOLD_BLACK", "BACKGROUND_YELLOW"),
})
```
2. Run the commands:
```
mkdir test1
chmod +t test1
touch test2
```
3. Run `ls -l`

4. Type `cat test2` without ENTER

- Problem 1.: `test2` is shown green in `ls`, but red on the `cat` prompt
- Problem 2.: (I don't have a test case yet) (edit: see edit5 below)
- Problem 3.: the yellow background of `test1` spills on the next lines.

## EDIT2:

Sorry, the story with `mi` and `or` was wrong, adding them to $LS_COLORS works.

## EDIT3:

The spilling only appears when `rs` is redefined in the $LS_COLORS, as above. This proves the point that it is a reset command, and as such `fi` should be used for files instead. So probably 3. is not an issue by itself, but merely a consequence of 1.

## EDIT4:

Sorry for the multiple edits, my understanding of the subject improved gradually while submitting this issue... :-)

## EDIT5:

I looked up `mh`. They simply represent inodes with multiple names. So for instance

```
touch test3
ln test3 test4
```
will create two links pointing to the same file, and `ls` will display them both using `mh` format.

To test this in python/xonsh, one can use `os.lstat` (or `os.stat`, depending of how one wants to handle symlinks):

```
$ import os
$ os.lstat("test3").st_nlink
2
```

**Comment by bobhy** [comment]
@qwenger, thanks for the thorough research!  
'fi' is a nasty suprise, it does not appear in my `dircolors` output!  (ubuntu 20.04) 
'mh' seems clear, given your work.  (it's mh=0 in my dircolors)

Any other changes you'd recommend?

**Comment by qwenger** [comment]
> @qwenger, thanks for the thorough research!

My pleasure.

> 'fi' is a nasty suprise, it does not appear in my `dircolors` output! (ubuntu 20.04)

Yeah it seems to vary between distros... my link [2] above doesn't mention it either...
Nevertheless I assume that your `ls` supports it, that is it reacts if you set `fi` in your `$LS_COLORS`, right?

> In some quick experiments with 'mh', it seems to be lower precedence than, for example, setXid and executable. Meaning the link to an executable file gets colored 'ex' not 'mh'. Does that sound right? (We're all learning from this issue!)

Yes it seems that `mh` has low precedence.

Maybe we can even ask directly? :-)
https://github.com/coreutils/coreutils/blob/master/src/ls.c#L4881

> Any other changes you'd recommend?

I don't see any for now, with my limited knowledge in the domain.

Looking at https://github.com/coreutils/coreutils/blob/master/src/ls.c#L597 it seems that my table is complete, except for `cl`, which is again a special formatting code.

The goal is of course that `color_file(...)` mimics `ls` colors (and their precedences) as close as possible - so `fi` seems quite necessary and `mh` is a nice to have if we can live with the penalty of an additional call to `(l)stat`.

In a rapid glance, `color_file` seems to follow `ls`'s `get_color_indicator`, but I haven't checked all branchings in detail.

Maybe we can also ask why `ca` is not supported? (a justification could be that is probably not as straightforward and more costly to implement as the rest - see https://github.com/coreutils/coreutils/blob/master/src/ls.c#L3123 and https://stackoverflow.com/questions/21560856/how-to-read-file-capabilities-using-python for hints)

**Comment by bobhy** [comment]
Good to hear this is mostly on track.

The only reason 'ca' isn't there is I didn't know how to test for it.  Thanks for the pointers, maybe I can figure it out now.  Also on the track of ln=target, but there's a color mapping issue there too, #3578.

**Comment by qwenger** [comment]
For `ca` it seems that it is only a matter of checking

```
$ import os
$ bool(os.listxattr("/bin/ping")
True
$ bool(os.listxattr("/bin/echo"))
False
```

`listxattr` is only available for Python >=3.3 on Linux, though.

**Comment by bobhy** [comment]
Per my man page, and [this article](https://linux-audit.com/using-xattrs-extended-attributes-on-linux/), capabilities are stored in the specific extended attribute 'security.capability', so I'll have to check for that attribute name (and hope there is not a localization concern).

This is a linux-only thing, apparently [not in BSD](https://unix.stackexchange.com/questions/503826/does-bsd-have-priveleges-or-capabilities) or Unix.  

Not on Windows, of course.
```
$ fs = os.stat('/usr/bin/ping')                                                                                         
$ stat.S_ISREG(fs.st_mode)                                                                                              
True
$ os.listxattr('/usr/bin/ping')                                                                                         
['security.capability']
$ getcap /usr/bin/ping                                                                                                  
/usr/bin/ping = cap_net_raw+ep
```
It is a regular file, so the second OS call will be frequent, but only after we know the file exists.  
`ls --color` uses the ca color for an executable file, so it has high precedence (unlike mh).

All in all, seems feasible.  But now I know why I left it till V2!

**Comment by qwenger** [comment]
> Per my man page, and [this article](https://linux-audit.com/using-xattrs-extended-attributes-on-linux/), capabilities are stored in the specific extended attribute 'security.capability', so I'll have to check for that attribute name (and hope there is not a localization concern).

Oh, right. Sorry I mixed up extended attributes and capabilities. So something like

```
"security.capability" in os.listxattr(path)
```

or

```
try:
    os.getxattr(path, "security.capability")
except (AttributeError, OSError):
    # AttributeError if no getxattr, OSError if no security cap
```

AFAICS there's no localization taking place, but I'm not sure.

> `ls --color` uses the ca color for an executable file, so it has high precedence (unlike mh).

Yes cap is between setXid and exec. https://github.com/coreutils/coreutils/blob/master/src/ls.c#L4922.

> All in all, seems feasible. But now I know why I left it till V2!

Again, `fi` is my main concern. But if we can get `ca` as well it's nice.

**Comment by bobhy** [comment]
PR is done (#3613), feel free to take a look.  Might be a while till this is approved, CI seems to be having problems at the moment (#3616)

**Comment by qwenger** [comment]
I had written two comments there already, but forgot to submit the review... done now.

**Comment by melund** [comment]
@qwenger The issue has been fixed by @bobhy in #3613. Thanks a ton for the bug report and the review.

**Comment by qwenger** [comment]
Many, many thanks @bobhy and everyone involved for your work!
